agent/consul/intention_endpoint_test.go

```diff
@@ -389,6 +389,325 @@ service "foo" {
+	})
+
+	doIntentionUpdate := func(t *testing.T, token string, dest string, deny bool) {
+		t.Helper()
+		requireAllowWrite(t)
+		ixn := structs.IntentionRequest{
+			Datacenter: "dc1",
+			Op:         structs.IntentionOpUpdate,
+			Intention: &structs.Intention{
+				ID:              intentionID,
+				SourceNS:        "default",
+				SourceName:      "*",
+				DestinationNS:   "default",
+				DestinationName: dest,
+				Action:          structs.IntentionActionAllow,
+				SourceType:      structs.IntentionSourceConsul,
+			},
+			WriteRequest: structs.WriteRequest{Token: token},
+		}
+		var reply string
+		err := msgpackrpc.CallWithCodec(codec, "Intention.Apply", &ixn, &reply)
+		if deny {
+			require.Error(t, err)
+			require.True(t, acl.IsErrPermissionDenied(err))
+		} else {
+			require.NoError(t, err)
+		}
+	}
+
+	t.Run("deny-update-for-foo-token", func(t *testing.T) {
+		doIntentionUpdate(t, fooToken.SecretID, "foo", true)
+	})
+
+	t.Run("allow-update-for-prefix-token", func(t *testing.T) {
+		// this tests that regardless of going from a wildcard intention
+		// to a non-wildcard or the opposite direction that the permissions
+		// are checked correctly. This also happens to leave the intention
+		// in a state ready for verifying similar things with deletion
+		doIntentionUpdate(t, writeToken.SecretID, "foo", false)
+		doIntentionUpdate(t, writeToken.SecretID, "*", false)
+	})
+
+	doIntentionDelete := func(t *testing.T, token string, deny bool) {
+		t.Helper()
+		requireAllowWrite(t)
+		ixn := structs.IntentionRequest{
+			Datacenter: "dc1",
+			Op:         structs.IntentionOpDelete,
+			Intention: &structs.Intention{
+				ID: intentionID,
+			},
+			WriteRequest: structs.WriteRequest{Token: token},
+		}
+		var reply string
+		err := msgpackrpc.CallWithCodec(codec, "Intention.Apply", &ixn, &reply)
+		if deny {
+			require.Error(t, err)
+			require.True(t, acl.IsErrPermissionDenied(err))
+		} else {
+			require.NoError(t, err)
+		}
+	}
+
+	t.Run("deny-delete-for-read-token", func(t *testing.T) {
+		doIntentionDelete(t, readToken.SecretID, true)
+	})
+
+	t.Run("deny-delete-for-exact-wildcard-rule", func(t *testing.T) {
+		// This test ensures that having a rules like:
+		// service "*" {
+		//    intentions = "write"
+		// }
+		// will not actually allow deleting an intention with a wildcard service name
+		doIntentionDelete(t, exactToken.SecretID, true)
+	})
+
+	t.Run("deny-delete-for-prefix-wildcard-rule", func(t *testing.T) {
+		// This test ensures that having a rules like:
+		// service_prefix "*" {
+		//    intentions = "write"
+		// }
+		// will not actually allow creating an intention with a wildcard service name
+		doIntentionDelete(t, wildcardPrefixToken.SecretID, true)
+	})
+
+	t.Run("allow-delete", func(t *testing.T) {
+		// tests that a token with all the required privileges can delete
+		// intentions with a wildcard destination
+		doIntentionDelete(t, writeToken.SecretID, false)
+	})
+}
+
```

Comment:
You can parallelize subtests (by calling `t.Parallel()` within the func passed to `t.Run`) but in general our team has found this to be difficult to maintain. In many cases it would be fine (and we have parallelized them) but in many others you would have to take extra care to ensure that any resources setup outside of the parallelized subtests live for as long as all the other tests which would involve wait groups and such for many of our tests (any of those that defer calls to shutdown a test agent or remove a directory).

So our rule of thumb in general is to not parallelize subtest invocations.

Additionally for these tests, the "subtest" is really just meant to separate different steps of a test to make pin pointing failures easier. In these tests I have also put in logic so that subtest dependencies are handled and we will skip subtests that require a previous subtest completely successfully.

This is quite odd, but in reality this test along with most of our test suite are really integration level tests. There is no way currently to test an RPC endpoint outside of spinning up a server running raft along with everything else. So this test could be broken into 15 other tests that start a server, wait for various conditions and then do their singular check and finish. However our test sweet becomes more flaky (and takes longer to execute as we spin up more test servers/agents). So writing the test in this manner: 1 test server + subtests + subtest skipping based on dependencies is to allow pinpointing problems when a test fails easier *and* preventing many test server startups to reduce testing time and prevent flakiness.